sirepo/auth_role_moderation.py

```diff
@@ -0,0 +1,172 @@
+        )
+
+    req = sirepo.http_request.parse_post(require_sim_type=False)
+    d = req.req_data
+    u = sirepo.auth.logged_in_user()
+    r = sirepo.auth_role.for_sim_type(d.simulationType)
+    with sirepo.util.THREAD_LOCK:
+        if sirepo.auth_db.UserRole.has_role(u, r):
+            raise sirepo.util.Redirect(sirepo.uri.local_route(d.simulationType))
+        try:
+            sirepo.auth_db.UserRoleInvite(
+                uid=u,
+                role=r,
+                status='pending',
+                token=sirepo.util.random_base62(32),
+            ).save()
+        except sqlalchemy.exc.IntegrityError as e:
+            pkdlog('Error={} saving UserRoleInvite for uid={} role={} stack={}', e, u, r, pkdexc())
+            raise sirepo.util.UserAlert(
+                f"You've already submitted a moderation request.",
+            )
+
+    l = _request_link(d.simulationType)
+    if sirepo.auth.dev_and_no_method('email'):
+        pkdlog('Email not enabled. Moderation reason={} Link to approve/deny={}', d.reason, l)
+    else:
+        _send_request_email(
+            PKDict(
+                display_name=sirepo.auth.user_display_name(u),
+                email_addr=sirepo.auth.user_name(),
+                link=l,
+                reason=d.reason,
+                role=sirepo.auth_role.for_sim_type(d.simulationType),
+                sim_type=d.simulationType,
+                uid=u,
+            ).pkupdate(sirepo.http_request.user_agent_headers())
+        )
+    return sirepo.http_reply.gen_json_ok()
+
+
+@sirepo.api_perm.require_adm
+def api_getPendingAndClarifyUserRoleInvites():
+    return sirepo.http_reply.gen_json(PKDict(
+        rows=[r.as_pkdict() for r in (
+                sirepo.auth_db.UserRoleInvite.get_pending_rows()
+                + sirepo.auth_db.UserRoleInvite.get_clarify_rows()
+        )],
+    ))
+
+
+@sirepo.api_perm.require_adm
+def api_approveOrDenyRole():
+    def _send_approve_or_deny_email(info):
+        a = 'Approved' if info.approved else 'Denied'
+        if sirepo.auth.dev_and_no_method('email'):
+            pkdlog('Email not enabled. Access is {}', a)
+            return
+        sirepo.smtp.send(
+            recipient=sirepo.auth.get_module(
+                'email',
+            ).AuthEmailUser.search_by(uid=info.uid).unverified_email,
+            subject=f'{info.app_name} Access Request ' + a,
+            body=pkjinja.render_resource(
+                f'auth_role_moderation/{info.status}_email',
+                PKDict(
+                    link=sirepo.uri.server_root(external=True) + sirepo.auth_role.sim_type(info.role),
+                    app_name=info.app_name,
+                ),
+            ),
+        )
+    req = sirepo.http_request.parse_post(require_sim_type=False, type=False)
+    i = sirepo.auth_db.UserRoleInvite.search_by(token=req.req_data.token)
+    assert i, f'No record for token={req.req_data.token}'
```

Comment:
If the record is not found, it should be a UserAlert. Below the record is deleted if approved.

However, in looking at this from an auditing perspective, we could retain the record of approvals, too. This would make the table larger, but no larger than the role table, which is accessed much more frequently.  Indeed, the person who took the last action should be recorded in the table, too, so expand the table to include moderator_uid.

Reply:
8ef331f5ed8d6d578701a4374d899b1b06f51658